### ISSUE TITLE
genetic interactions (Pombase)

### ISSUE THREAD

**Issue by ValWood** [pombase/canto#2301]
Interface for curating genetic interaction.

- [x] ~We would probably like to be able to "qualify" "genetic interaction type"  during multi-allele annotation, rather than a separate annotation type. This  would 
i. streamline the annotation workflow so people are less likely to omit one of the 2 annotation types. ii. Force the  phenotypes annotations which are components of the genetic interaction to be made explicitly therefore make the annotation more complete. We are not yet sure how exactly it would work but brainstorming got this far: 1. Create a multi-gene phenotype annotation 2. Optional extra action to classify this multi-allele genotype as a particular type of genetic interaction (synthetic lethal etc)~  (See point 4& 5. below)

**Comment by kimrutherford** [comment]
I forgot about this issue.  This needs to be decided and implemented before the pombe Canto is changed to the new version of the code which does the genotype-genotype interactions.

So there's no need for me to shut down Canto over the weekend.

See also: https://github.com/pombase/pombase-chado/issues/739

**Comment by ValWood** [comment]
we can discuss this on next call.

**Comment by ValWood** [comment]
I think this is everything we discussed yesterday: Please read and edit if I got anything incorrect.
@JackyVH 
@juanmatacambridge
@kimrutherford 
@mah11 

1. Because whether a phenotype is required will depend on interaction type we need to  add genotypes and interaction type before allowing phenotype selection. Once interaction type selected:
2. Display  relevant existing phenotypes  (and their conditions) for the relevant single allele will be presented in the dropdown.
3. Provide an option  to add a phenotype and conditions if not already curated (this would also populate the single allele phenotype table).
4. Later: Add  prompts if there is no  phenotype for the  multi allele combination (and  maybe even capture it here)
5. Later: Once the GI interface is in place, look into the possibility of having a "quick route" to create GI's for 2-gene interactions from the multi-gene phenotype route. We might be able to automatically create some of these. For example, if the user created  2 viable single gene genotypes and an enviable multi for the same allele combination we could automagically create the "synthetic lethal" GI. 


Interaction type  |  Phenotype to report |  FYPO branch restriction for phenotype acted on | Notes
--- | --- | --- | ---
Synthetic haploinsufficiency | omit, unused, keep it simple can add later if required |
Dosage growth defect | enhanced growth phenotype (genotype1) |  FYPO:0000046 decreased cell population growth |
Dosage rescue | rescued phenotype (genotype1) |  no term restriction | 
Phenotypic enhancement | enhanced  phenotype (genotype1) | no term restriction | 
Phenotypic suppression  | suppressed phenotype (genotype1) |  no term restriction | 
Synthetic rescue | rescued phenotype (inviability or growth defect) (genotype1) | FYPO:0000046 decreased cell population growth OR FYPO:0002061 inviable vegetative cell population | prompt if no multi-allele  annotation  for allele combination in FYPO:0002060 viable vegetative cell population OR FYPO:0003937 increased cell population growth
Synthetic growth defect | outcome is compromised growth, do not need to record here? | n/a | prompt if no multi-allele  annotation  for allele combination in  FYPO:0000046 decreased cell population growth branch | 
Synthetic lethality  | outcome is inviability, do not need to record here | n/a | prompt if no multi-allele  annotation   for allele combination in FYPO:0002061 inviable vegetative cell population  branch
Dosage lethality |  outcome is inviability, do not need to record here | n/a | prompt if no multi-allele  annotation  for allele combination in FYPO:0002061 inviable vegetative cell population branch

**Comment by ValWood** [comment]
Also might need to consider extensions in some instances (or allow editing of penetrance/severity qualifiers)

**Comment by kimrutherford** [comment]
> "prompt if no multi-allele annotation"

Can you explain this a bit more?

**Comment by kimrutherford** [comment]
Changing the code to force selecting the interaction type first, then the phenotype is going to be painful.  It's the exact opposite of how it currently works for FlyBase so the code will need to support both options.  I'll put my thinking head on.

> 2. Display relevant existing phenotypes (and their conditions) for the relevant single allele will be presented in the dropdown.
> 3. Provide an option to add a phenotype and conditions if not already curated (this would also populate the single allele phenotype table).

This sound like how things are working already.  Are any changes needed?

**Comment by ValWood** [comment]
> "prompt if no multi-allele annotation"

This can happen later.

if the user selects, for example 

cdc13-117 (C379Y aa) cdc2-4w (C67F aa) as "synthetic lethal"

There should be an existing multi-allele annotation 

under  FYPO:0002061

as there is in this example

https://www.pombase.org/genotype/cdc13-117-C379Y-amino_acid_mutation-expression-not_assayed__cdc2-4w-C67F-amino_acid_mutation-expression-not_assayed

**Comment by ValWood** [comment]
> Changing the code to force selecting the interaction type first, then the phenotype is going to be painful. It's the exact opposite of how it currently works for FlyBase so the code will need to support both options. I'll put my thinking head on.

I was wondering if we could get away without the need to change the workflow.
But it's going to be very confusing to see the phenotype option available, and then for it not to be required if you select specific interaction types.  Also   "genotype 1" would not mean anything for the non-directional interactions,  so we couldn't just collect it anyway...

**Comment by ValWood** [comment]
>Display relevant existing phenotypes (and their conditions) for the relevant single allele will be presented in the dropdown.

I don't think this part is working?

In this e.g. (test/curs/8c7b04a443ae9d43/feature/genotype/view/1)
prp16-F528S(aaF528S)[Not assayed] | Synthetic Growth Defect | prp8-1(unknown)[Not assayed] / unspecified-cdc28(unspecified) unspecified-prp22(unspecified)

I don't see these phenotypes prompted:
prp16 | prp16-F528S(aaF528S)[Not assayed] | FYPO:0000080 | decreased cell population growth at low temperature | Cell growth assay | glucose MM, low temperature, agar plates |  
prp16 | prp16-F528S(aaF528S)[Not assayed] | FYPO:0000091 | sensitive to thiabendazole | Cell growth assay | glucose MM |   
prp16 | prp16-F528S(aaF528S)[Not assayed] | FYPO:0000082 | decreased cell population growth at high temperature | Cell growth assay | glucose MM, agar plates

**Comment by kimrutherford** [comment]
> I don't think this part is working?

Sorry, I needed to add it to the configuration for the test Canto instance.  It should be working now.

**Comment by ValWood** [comment]
Great. I  now see the annotated phenotypes.
- [ ] We now need to work on the restrictions to only display the phenotypes in table 3 above,
- [ ] Also remove redundancy where phenotypes are listed twice (or if these are conditions display the conditions too, maybe  [ conditions: standard temperature, glucose MM])

We can probably bring in some more restrictions. Possible suggestions if the list is still sometimes  too long (but could be added later):

 - [ ] In the non-growth phenotypes, we might be able to omit "normal" phenotypes for a subset? (possibly not all?)
-  [ ]  Probably for the phenotypes with "assayed using" these are unlikely to be used in GIs. We could therefore maybe collapse them to show the phenotype only, without the extension.

These changes should get the list a more manageable size to scan quickly. We need to aim to get rid of everything that can't be relevant.


<img width="816" alt="Screenshot 2021-01-25 at 12 55 37" src="https://user-images.githubusercontent.com/7359272/105708711-ba9c8c00-5f0c-11eb-8103-ba78149a6528.png">

**Comment by kimrutherford** [comment]
We should probably have a think about how these changes fit into the "long workflow".  Currently we expect new users to (initially) make phenotype annotations using the long workflow.  But there is no long workflow for interactions.  Is that a problem?

**Comment by kimrutherford** [comment]
We talked about adding the interaction details when making the genotype/phenotype annotation.  And I offered to make a mock up as a talking point.  I'm not sure it's useful but here it is:

![genotype-interactions-mock-up-1](https://user-images.githubusercontent.com/90474/109124159-7c1a1b80-77af-11eb-9634-ee7c469f451a.png)

**Comment by ValWood** [comment]
Spoke with Kim.

We would like to proceed with the ability to create a GI from an exiting multi-allele phenotype annotation.
The symmetrical cases are relatively straightforward.

For none symmetrical:

Example multi-allele annotation
Delta-zfs1 delta-puc1
Normal mating efficiency

Would have an option to add a GI

1. First an option to select interaction type would be presented
e.g. Option [ Phenotypic suppression ]
              
2. The  correct ordering would be selected based on the relation
e.g.
Delta-zfs1  rescued by  Delta-puc1  [x]
Delta-puc1  rescued by Delta-zfs1

3. All phenotypes of this allele  (with conditions) would be presented

e.g
FYPO:0000708 decreased mating efficiency low nitrogen glucose MM [x]
User might want to select more than one

Q. question, should the conditions be the same for the rescued/suppressed genotype?

4. After this is implemented we can add restrictions to phenotype terms displayed

We discussed the storage and display a little Kim will continue to think about this.
Most likely all existing alleles will be labelled as 'legacy'. They will no longer be able to be edited, but can be deleted and recreated in the new GI system.
Problems for Kim are storage because we want to treat as independent annotations in the display, but they will be coupled when curated. 

The GI changes will have the advantage of 
- Making the  GI's and multi alleles much more aligned (consistent)- i.e. reducing ommissions where either only  GI or multi-gene are made
- Make it faster to curate
- Make the data more useful as coupled to alleles and phenotypes.

We still need to discuss:
How changes will fit into the long workflow
We will still want to host 'legacy type' from BioGRID

**Comment by kimrutherford** [comment]
An advantage of this plan is that we can go ahead with updating PomBase Canto to use the same code as PHI-Canto and Fly-Canto.  That version of Canto still has the code to edit gene-gene interactions so we can update and keep the interface more or less as it is in the PomBase version now.

Once the new interface is implemented we can disable the gene-gene interaction editing.

I think the next step will be to copy the current annotation across to the test Canto, and configure the gene-gene interaction editing so we can check that all is OK.  I'll do that after we chat about this plan on Thursday.

**Comment by kimrutherford** [comment]
I've done a bit more thinking and I wonder about cases like this:
https://www.pombase.org/reference/PMID:24344203
where tsc2 and isp7 interact with multiple different interaction types.

Is that going to be a problem for the new interface?

I think the interactions for that paper are mostly from BioGRID and the notes are mostly different for each case.

**Comment by ValWood** [comment]
Good point, but I think it is OK. These must be different allele combinations. The same allele type can't be more than one of these  GI classifiers. However, a specific allele of one gene combined with different alleles of another gene is often done to show different interaction types. i.e. overexpression and deletion will usually have opposite effects.

**Comment by kimrutherford** [comment]
> but I think it is OK

That's good news.

My follow up question is: can genotype A rescue the function of genotype B at the same time as B rescues the function of A?

**Comment by ValWood** [comment]
> My follow up question is: can genotype A rescue the function of genotype B at the same time as B rescues the function of A?

Yes, theoretically (and I'm sure we have some examples).

Say both delta-1 and delta-b grow poorly but. combined they grow better. That would be a "synthetic reuse" of both. 
I think we discussed this and it would. be OK  in this case to make 2 annotations, one with each rescue. However, this would result in duplicate muti-gene annotation.

The other option would be to be able to select BOTH A. rescues B and B rescues A. This might be better?

**Comment by kimrutherford** [comment]
Could A rescue a phenotype of B and B suppress a phenotype of A at the same time?

**Comment by ValWood** [comment]
I think the evidences are constructed  so they are supposedly mutually exclusive:

Dosage Rescue - Overexpression/increased dosage of one gene rescues the lethality or growth defect of a strain mutated/deleted for another gene 

Phenotypic Suppression - Mutation/deletion/overexpression of one gene results in suppression of any phenotype (**other than lethality/growth defect)** associated with mutation/deletion/overexpression of another gene 

based on the **type** of phenotype. But I know there are some when I need to check which one to use. I think some are a bit fuzzy, but we would only use one.

So dosage rescue is phenotypic suppression for growth and viability. Phenotypic suppression is a mixed bag because it includes rescue by deleting and by overexpressing which have completely different interpretations for reconstructing a pathway!

**Comment by kimrutherford** [comment]
I've spent the week (off and on) thinking about how to implement this.  There are no good options and I'm not yet sure which is the least bad option.

So I'm going to put it aside for a little while and concentrate on moving the main pombe Canto over to the same code as FlyCanto and PHI-Canto.

**Comment by kimrutherford** [comment]
> So I'm going to put it aside for a little while and concentrate on moving the main pombe Canto over to the same code as FlyCanto and PHI-Canto.

As first step, I've copied the main Canto database over to test instance and updated it to use the latest version of the new code (used by FlyCanto and PHI-Canto).  I've had a quick look and everything seems mostly OK.  There have been some minor changes to how things look, like table columns getting slightly wider or narrower.  Hopefully nothing too annoying.

I'd appreciate if you both could have a quick look at some point and test making a few annotations.  There have been a lot of changes to the phenotype/genotype code under the hood but everything is still supposed to work the same way.

In a way most of the changes have been tested pretty well because Val has been using PHI-Canto which is using this new code.   PHI-Canto has a few extra options turned on (like pathogen host interactions) and a few turned off so testing the test instance would still be helpful.

The most likely problems you'll so are cases where functionality tweaks due the PHI-Canto and FlyCanto options have leaked through into the pombe version.  They should be quick fixes.  I've spotted one already: https://github.com/pombase/canto/issues/2434

We should chat about this on Thursday too.

**Comment by ValWood** [comment]
I haven't looked at this yet. I hope to get to it today.

**Comment by kimrutherford** [comment]
> We should chat about this on Thursday too.

Sounds like we won't have time to chat about this on Thursday.

**Comment by mah11** [comment]
I've just done a really quick check, and so far it looks OK except for one thing: the pulldown of interaction types doesn't include "Synthetic Lethality" (but it does have "Synthetic Haploinsufficiency"; did the wrong one get the ax in https://github.com/pombase/pombase-config/issues/60?).

**Comment by kimrutherford** [comment]
Yes, sorry about that.  Thanks for checking.  Hopefully I've removed the correct interaction type this time.

**Comment by mah11** [comment]
Yep; thanks!

**Comment by ValWood** [comment]
I have done quite extensive testing.

adding genes, genotypes single/multi/diploid, copying and editing genotypes, all curation types, GO/Modification/genetic/physical interaction. Copy edit annotation. Quicklinks, long workflow.

I  haven't found any problems.

**Comment by mah11** [comment]
It looks like Val has been more thorough than I have, but I've also seen no problems.

**Comment by kimrutherford** [comment]
Thanks for testing.  I think we should swap to the new code one day when there's not too much community curation going on.  We'll need 30 minutes of downtime.  There's no rush so I might do it next weekend.

**Comment by kimrutherford** [comment]
I'm getting there on this issue.

It's now possible to:
  - make interactions in the phenotype annotation edit dialog
  - save the annotation with "OK", which now stores the interaction(s) in the database
  - come back later and see the interaction(s) in the dialog:

![genotype-interactions-db-1](https://user-images.githubusercontent.com/90474/114683274-ff91cb00-9d63-11eb-8efd-07e8d41bcbd3.png)

------------------------------

Still to do  (not a complete list):
  - [ ] add code for viewing phenotypes of the interactions
  - [ ] fix database storage code so that interactions can be deleted as well as deleted
  - [ ] show a table of genotype-genotype interactions on the front page (currently they're only visible in the edit dialog)

**Comment by kimrutherford** [comment]
These two items are done now:

>  -  add code for viewing phenotypes of the interactions
>  -  fix database storage code so that interactions can be deleted as well as created

**Comment by kimrutherford** [comment]
The pombe Canto is now using the latest code.  Everything seems OK to me but please let me know if there are any problems.

**Comment by kimrutherford** [comment]
> The pombe Canto is now using the latest code. Everything seems OK to me but please let me know if there are any problems.

The test Canto is using the latest code too with the addition of the genotype interaction changes.

**Comment by ValWood** [comment]
>The pombe Canto is now using the latest code. Everything seems OK to me but please let me know if there are any problems.

I nearly finished curating a paper, and everything seems good so far.
No problems so far adding gene, genotypes (single/multi), GO terms, phenotypes, modifications, extensions .....

Also did a couple of edits to older sessions and that went OK too.

**Comment by ValWood** [comment]
The only thing I notice is that the genes don't show for existing genetic interactions, but maybe that was happening right now anyway?

**Comment by mah11** [comment]
I think Val has tested more thoroughly than I have, but I'm not having any problems (other than the existing genetic interactions thing) either.

**Comment by ValWood** [comment]
OK I finished testing. I did pretty much everything I would do  during normal paper curation, so I suspect if any problems they are very minor...

**Comment by kimrutherford** [comment]
Thanks you both for testing.  The one big job done.

**Comment by mah11** [comment]
I did a little more testing on the new genetic interaction business in test canto ...

I'm still getting that bug we saw in your demo, where it only shows one phenotype even if there are more. Now it only happens if I finish the annotation and then go back to edit it (using the Edit link in the table). If I make an interaction and choose two phenotypes, then view again before I finish the whole phenotype annotation, I see both. If this is confusing I'm happy to demo.

**Comment by mah11** [comment]
One more thing I forgot to say yesterday: is there a plan to add a way to see which phenotype annotations have interactions associated with them?

**Comment by kimrutherford** [comment]
> I'm still getting that bug we saw in your demo, where it only shows one phenotype even if there are more.

Thanks, I'd forgotten about that.  I'll fix it next week.

**Comment by kimrutherford** [comment]
> One more thing I forgot to say yesterday: is there a plan to add a way to see which phenotype annotations have interactions associated with them?

That sounds like a good idea.  We'll need to chat about where and how to show this.

**Comment by kimrutherford** [comment]
> I'm still getting that bug we saw in your demo, where it only shows one phenotype even if there are more.

That's fixed now.

**Comment by kimrutherford** [comment]
> The only thing I notice is that the genes don't show for existing genetic interactions, but maybe that was happening right now anyway?

That still needs fixing.  I'm commenting so it doesn't get forgotten in the discussion.

**Comment by kimrutherford** [comment]
> The only thing I notice is that the genes don't show for existing genetic interactions, but maybe that was happening right now anyway?

Thanks.  I've fixed that now.

**Comment by ValWood** [comment]
This ticket has 48 comments.
We should close off everything that isn't relevant and put any outstanding issues in new tickets.

**Comment by ValWood** [comment]
I should go through this ticket and open a new ticket for any outstanding tasks...

**Comment by ValWood** [comment]
I think everything outstanding in this ticket is covered by https://github.com/pombase/canto/issues/2551

**Comment by ValWood** [comment]
could you glance over?
There is the issue of the long interface, but I think we can discuss  this later with a view to  using only the quick interface with some. enhancements.

**Comment by kimrutherford** [comment]
> I think everything outstanding in this ticket is covered by https://github.com/pombase/canto/issues/2551

I think so too.
